### ISSUE TITLE
Fix cctest/test-simplified-lowering/RunAccessTests_float64

### ISSUE THREAD

**Issue by paul99** [v8mips/v8mips#98]
This test fails on popcorn board in release mode. And passes in debug mode. It does _not_ fail on er8. Whatever makes it crash does not crash in FP kernel emulation.

**Comment by paul99** [comment]
Balázs has investigated this in detail, and has a trial patch to fix it. I will move the discussion from slack to here.

kilvadyb [9:38 AM] 
The problem is that that the test uses Handle<ByteArray> tagged_array;  // TODO(titzer): use FixedArray for tagged.

kilvadyb [9:40 AM]
For double values it fails as the address of the allocated ByteArray is not 8 bytes aligned. I am working on to add FixedArray usage as the comment suggests.

kilvadyb [9:42 AM]
But I can not find a nice way yet. Actually FixedTypedArray<> should by used as that class has proper get/set functions which are implemented with our FIELD_DOUBLE_READ/_WRITE macros.

paul [9:47 AM] 
Hi Balázs - I would worry about fixed typed array, as typed arrays are a specific construct within JS, that is not the same as array. I don't know for a fact that this would be incorrect, but I am very suspicious.

paul [9:47 AM]
I guess we could look for uses for FixedTypedArray within v8 and the test suites to see if it is used elsewhere. (meaning other than in the implementation and testing of the JS typed array feature) (edited)

paul [9:48 AM]
If you use FixedArray, I guess that has only 4-byte alignment.

paul [9:48 AM]
AnotherQ: these tests pass on sim, but I thought we generated error for unaligned doubles in the sim. (edited)

kilvadyb [9:49 AM] 
I can not use FixedArray yet. The problematic cctest helper class is:

kilvadyb [9:49 AM]

```
// A helper class for accessing fields and elements of various types, on both
// tagged and untagged base pointers. Contains both tagged and untagged buffers
// for testing direct memory access from generated code.
template <typename E>
class AccessTester : public HandleAndZoneScope {
```

paul [9:51 AM] 
Maybe we will need add a mips32 unaligned accessor there, as we did in objects-inl.h

kilvadyb [9:51 AM] 
I have not checked why it is ok on sim. It passes in debug mode on the boards also al ldc1 is not used in debug mode.

kilvadyb [9:54 AM]
That was my first idea but if we could use FixedTypedArrays then our align-fix implementations from object-inl.h would be used automatically.

kilvadyb [9:56 AM]
So if I could use a proper FixedTypedArray<> based on the E type of AccessTester<E> then it would be absolutely fine. But if they want raw memory access then we should add our aligmen macroes to cctest/compiler/test-simplified-lowering.cc also.

paul [9:58 AM] 
I'm not familiar enough with this code to comment in detail. I will look at it a bit later in the morning. I need to leave to go to office now. Back online in about an hour.

paul [10:00 AM]
Meta question: should we create a channel for this bug, since we have a long conversation here?Akos message above about other bugs is long scrolled off my screen (its the thread message issue, and we before thought to create short-lived-channels - that could be used, and then archived once we had (for instance) fixed the bug)

paul [10:02 AM]
And finally: I did not understand this: "It passes in debug mode on the boards also al ldc1 is not used in debug mode." Did you look at the code, to see that ldc1 is not being used, or are you supposing that it is not, since it does not error? (it could be in debug mode things are aligned for some reason?)

kilvadyb [10:06 AM] 
yes, new channel is a good idea. No I have not checked the assembly code in debug mode but all my earlier debugs showed thet O(0, 1, 2)  are ok and O(3) == release mode causes the aligment problem with ldc1.

paul [10:09 AM] 
very interesting...

kilvadyb [10:18 AM] 
I check in debug mode and try an alternate template impl. for doubles. I will report the results soon.

kilvadyb [10:51 AM] 
In release mode:
0x5016e8 <_ZL26TestRunAccessTests_float64v+796> ldc1   $f0,7(v0)
v0: 0x54009485 + 7 == 5400948c

In debug (offsets are less as lot of function calls are not inlined): 
0xa44bf0 <_ZL26TestRunAccessTests_float64v+124> mfc1   v0,$f0  
0xa44bf4 <_ZL26TestRunAccessTests_float64v+128> mfc1   v1,$f1

kilvadyb [12:49 PM] 
The following patch fixes the test-simplified-lowering/RunAccessTests_float64 test although seems to be too much to me

The patch is in slack here: https://v8m.slack.com/files/kilvadyb/F02R45K24/tmp.diff

**Comment by paul99** [comment]
I agree that this patch is a little heavy-weight, but it does have the nice advantage of working. :+1:  However, I'd like to see if we can find something more minimal.

Above there was discussion of this test passing on sim, which is silly of course, since there is no code-generation and the sim is never run. This is all C++ code. Only way to test for mips is to test on mips.

Balázs answered the debug/release question, gcc generates ldc1 only on O3, and falls back to pairs of lwc1 for lower opt levels.

The question is how to either get all this aligned for doubles, or use accessor like the READ_DOUBLE_FIELD() in objects-inl.h.

It looks like FixedArray's have a header that is multiple of 8, so if we can coerce it to be aligned to start with, it should be OK. But allocating it thru the factory method has no parameter/method to specify alignment.

Balázs, I do see now why you were interested in using FixedTypedArray, as it does have the handling we want for doubles. Why don't you give that version a try, and let's see how it goes.

**Comment by paul99** [comment]
I just set this to pri-2, since we now understand what is happening, and this failure is a result of the way this unit-test is constructed, and does not represent a failure mode that would happen in real v8 code - that I can see. (if someone thinks otherwise, please let us know)

So we might wish to triage other bugs, which may be more critical, before spending too much more time on this one.

**Comment by kilvadyb** [comment]
I spent a lot of hours to find a simple way to use FixedTypedArray for doubles (FixedTypedArray<Float64ArrayTraits>). As AccessTester<E> is used for such types as well which does not have a matching ArrayTraits type/class (like Smi*, int64_t) The whole AccessTester<E> can not be modified to use FixedTypedArray<> instead of ByteArray. If I try to do a special AccessTesterDouble for doubles then I have to override the same amount of functions then in the first proposal patch. So after a lot of reading and trying I can't see a smaller/lighter solution then my first patch. It can be slightly shorter if I use "this->" instead of "AccessBase<E>::". That is the only improvement that I reached (in c++14 there is a new "using TemplateBase<T>" directive but as I read it is for constructors only and v8 is not using c++14 anyway).
